### PR TITLE
curl: rebuild when libopenssl config changes

### PR DESCRIPTION
**Pull request by dengqf6**
Commit 450d44a8ead2217f8acf541a4eaa4ad560b3e5ac disabled OpenSSL NPN by default, but libcurl was built with OpenSSL NPN.
This result in undefined reference to `SSL_CTX_set_next_protos_advertised_cb` error if libcurl is previously built against OpenSSL.

Add `CONFIG_OPENSSL_WITH_NPN` to `PKG_CONFIG_DEPENDS` so it will trigger rebuild everytime the option is changed.

### CODE REVIEW

**Comment by chunkeey** [comment]
> if libcurl is previously built against OpenSSL

I don't think ".. if curl is previously built ..." binaries is something that is supported by OpenWrt in that sense. Because for new image and release, you should refresh the config and recompile the libraries and binaries (make clean) so everything can fit together.

That said, I'm not 100% certain that I understood everything here.

**Comment by dengqf6** [comment]
See openwrt/packages#8730
I know `make clean` and rebuild will (always) work. But we'd better fix the underlying cause if possible.

**Comment by neheb** [comment]
ACK

**Comment by ynezz** [comment]
@cotequeiroz Could you please take look at this? Thanks.

File: package/network/utils/curl/Makefile [hunk 1]
```diff
@@ -64,7 +64,9 @@ PKG_CONFIG_DEPENDS:= \
   CONFIG_LIBCURL_TLS_SRP \
   CONFIG_LIBCURL_UNIX_SOCKETS \
   CONFIG_LIBCURL_VERBOSE \
-  CONFIG_LIBCURL_NTLM
+  CONFIG_LIBCURL_NTLM \
+  \
+  $(if $(CONFIG_LIBCURL_OPENSSL),CONFIG_OPENSSL_WITH_NPN,)
```
Comment (cotequeiroz) [inline package/network/utils/curl/Makefile]:
```suggestion
  $(if $(CONFIG_LIBCURL_OPENSSL), \
	CONFIG_OPENSSL_ENGINE \
	CONFIG_OPENSSL_WITH_COMPRESSION \
	CONFIG_OPENSSL_WITH_NPN \
	 \
	 \
)
```
Apparently, NPN is not the only configuration option being used by curl.  I've found at least these two in `lib/vtls/openssl.c`, and haven't checked what happens if `OPENSSL_WITH_EC` changes.

**Review by cotequeiroz — COMMENTED** [review]
I like your idea, but I would change your description.  You're not proposing a fix, but an enhancement.

The error you're trying to fix only happens in an unsupported condition.  You can't expect to change a feature option in a library package, and expect packages to adjust without rebuilding.  That's probably why @chunkeey didn't understand fully what you were trying to do.  It works the other way too.  If the package is compiled with NPN off, you can't expect it to use NPN by enabling it in openssl and rebuilding just openssl.  So this has little to do with 450d44a, it would have happened before it was applied.
What you are doing is adding a trigger to rebuild curl automatically, which is a good idea.
I'm suggesting to change the description to something like:
```
curl: rebuild when libopenssl config changes

When some libopenssl options change curl will have to be rebuild to
adapt to those changes, avoiding undefined reference errors or features
disabled in curl.

Add CONFIG_OPENSSL_WITH_NPN ... to PKG_CONFIG_DEPENDS so it will trigger
rebuild every time the options are changed.
```

**Comment by cotequeiroz** [comment]
> 
> 
> See [openwrt/packages#8730](https://github.com/openwrt/packages/pull/8730)
> I know `make clean` and rebuild will (always) work. But we'd better fix the underlying cause if possible.

openwrt/packages#8730 was somewhat a different case.  `OPENSSL_ENGINE` was off by default by the package Makefile, but it was always on when built by the bots, so the official repositories always had it on.  What happened was that people were building custom images, which had engines off by default, and then were installing packages from the official repos, so they would not find those engine symbols, just as you are reporting here for NPN (they do have similarities :-).    NPN is not being selected by any package, so the bots build the library and all dependent packages without NPN.  They won't break anything when installed in a custom image.  They won't be able to use NPN either, but that's part of mixing up custom-built firmware with standard-built packages.

**Comment by neheb** [comment]
This should be OK to merge.

**Comment by adrianschmutzler** [comment]
@dengqf6 Can you apply the changes requested by @cotequeiroz ?

**Comment by ldir-EDB0** [comment]
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ldir.git

**Pull request closed without merging**